{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kallisto, version 0.46.2\n"
     ]
    }
   ],
   "source": [
    "!kallisto version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since data is on external hard drive, have to mount it in wsl. Note this is a one time step for your machine.\n",
    "\n",
    "1. Open WSL\n",
    "    - Open Windows Terminal or Powershell and type 'wsl'\n",
    "2. Create a Mount Point\n",
    "    - Decide where you want to mount your network drive in your WSL file system\n",
    "    - Common choice is under /mnt/, like '/mnt/d'\n",
    "    - Create mount point by typing 'mkdir -p /mnt/d' in the terminal and press enter\n",
    "3. Mount the Network Drive\n",
    "    - Type 'sudo mount -t drvfs '\\\\Book-m08q8a5m4r\\d' /mnt/d' in the terminal and press enter.\n",
    "4. Verify the Mount\n",
    "    - Type 'ls /mnt/d' in the terminal and press enter\n",
    "    - If you see the contents then it was successful\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[build] loading fasta file /mnt/e/fastq/GCF_000001405.40/ncbi_dataset/data/GCF_000001405.40/rna.fna\n",
      "[build] k-mer length: 31\n",
      "[build] warning: clipped off poly-A tail (longer than 10)\n",
      "        from 5074 target sequences\n",
      "[build] warning: replaced 16 non-ACGUT characters in the input sequence\n",
      "        with pseudorandom nucleotides\n",
      "[build] counting k-mers ... done.\n",
      "[build] building target de Bruijn graph ...  done \n",
      "[build] creating equivalence classes ...  done\n",
      "[build] target de Bruijn graph has 999245 contigs and contains 142403527 k-mers \n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Took 8 1/2 minutes to run\n",
    "!kallisto index -i \"/mnt/e/fastq/GCF_000001405.40/ncbi_dataset/data/GCF_000001405.40/output_indexed.idx\" \"/mnt/e/fastq/GCF_000001405.40/ncbi_dataset/data/GCF_000001405.40/rna.fna\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0001_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0001_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 207,412,414 reads, 107,409,028 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 188.166\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,484 rounds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Took 93 minutes to run\n",
    "!kallisto quant -i \"/mnt/e/fastq/GCF_000001405.40/ncbi_dataset/data/GCF_000001405.40/output_indexed.idx\" -o \"/mnt/e/kallisto\" \"/mnt/e/20231030 - 10051-NM/10051-NM-0001_S1_L005_R1_001.fastq.gz\" \"/mnt/e/20231030 - 10051-NM/10051-NM-0001_S1_L005_R2_001.fastq.gz\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting pandas\n",
      "  Downloading pandas-2.1.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.3 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m12.3/12.3 MB\u001b[0m \u001b[31m13.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hCollecting pytz>=2020.1\n",
      "  Downloading pytz-2023.3.post1-py2.py3-none-any.whl (502 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m502.5/502.5 KB\u001b[0m \u001b[31m38.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hCollecting numpy<2,>=1.22.4\n",
      "  Downloading numpy-1.26.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.2 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m18.2/18.2 MB\u001b[0m \u001b[31m18.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hCollecting tzdata>=2022.1\n",
      "  Downloading tzdata-2023.3-py2.py3-none-any.whl (341 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m341.8/341.8 KB\u001b[0m \u001b[31m20.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: python-dateutil>=2.8.2 in /home/nimuser/.local/lib/python3.10/site-packages (from pandas) (2.8.2)\n",
      "Requirement already satisfied: six>=1.5 in /usr/lib/python3/dist-packages (from python-dateutil>=2.8.2->pandas) (1.16.0)\n",
      "Installing collected packages: pytz, tzdata, numpy, pandas\n",
      "\u001b[33m  WARNING: The script f2py is installed in '/home/nimuser/.local/bin' which is not on PATH.\n",
      "  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.\u001b[0m\u001b[33m\n",
      "\u001b[0mSuccessfully installed numpy-1.26.2 pandas-2.1.3 pytz-2023.3.post1 tzdata-2023.3\n"
     ]
    }
   ],
   "source": [
    "!pip install pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_id</th>\n",
       "      <th>length</th>\n",
       "      <th>eff_length</th>\n",
       "      <th>est_counts</th>\n",
       "      <th>tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_000014.6</td>\n",
       "      <td>4610</td>\n",
       "      <td>4422.83</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_000015.3</td>\n",
       "      <td>1285</td>\n",
       "      <td>1097.83</td>\n",
       "      <td>53.00</td>\n",
       "      <td>0.276341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_000016.6</td>\n",
       "      <td>2261</td>\n",
       "      <td>2073.83</td>\n",
       "      <td>1111.53</td>\n",
       "      <td>3.067990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_000017.4</td>\n",
       "      <td>1859</td>\n",
       "      <td>1671.83</td>\n",
       "      <td>2766.93</td>\n",
       "      <td>9.473520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_000018.4</td>\n",
       "      <td>2184</td>\n",
       "      <td>1996.83</td>\n",
       "      <td>31658.60</td>\n",
       "      <td>90.751800</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     target_id  length  eff_length  est_counts        tpm\n",
       "0  NM_000014.6    4610     4422.83        0.00   0.000000\n",
       "1  NM_000015.3    1285     1097.83       53.00   0.276341\n",
       "2  NM_000016.6    2261     2073.83     1111.53   3.067990\n",
       "3  NM_000017.4    1859     1671.83     2766.93   9.473520\n",
       "4  NM_000018.4    2184     1996.83    31658.60  90.751800"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "sample_1_df = pd.read_csv(\"/mnt/e/kallisto/abundance.tsv\", sep='\\t')\n",
    "sample_1_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_10897/929896097.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_counts.rename(columns={'tpm': 'Sample_1'}, inplace=True)  # or 'est_counts' as per your data\n"
     ]
    }
   ],
   "source": [
    "gene_counts = sample_1_df[['target_id', 'tpm']]  # or 'est_counts' depending on your need\n",
    "gene_counts.set_index('target_id', inplace=True)\n",
    "gene_counts.rename(columns={'tpm': 'Sample_1'}, inplace=True)  # or 'est_counts' as per your data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>target_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_000014.6</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000015.3</th>\n",
       "      <td>0.276341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000016.6</th>\n",
       "      <td>3.067990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000017.4</th>\n",
       "      <td>9.473520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000018.4</th>\n",
       "      <td>90.751800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XR_953251.3</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XR_953252.3</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XR_953253.3</th>\n",
       "      <td>0.098578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XR_953307.2</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XR_953308.2</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>185121 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Sample_1\n",
       "target_id             \n",
       "NM_000014.6   0.000000\n",
       "NM_000015.3   0.276341\n",
       "NM_000016.6   3.067990\n",
       "NM_000017.4   9.473520\n",
       "NM_000018.4  90.751800\n",
       "...                ...\n",
       "XR_953251.3   0.000000\n",
       "XR_953252.3   0.000000\n",
       "XR_953253.3   0.098578\n",
       "XR_953307.2   0.000000\n",
       "XR_953308.2   0.000000\n",
       "\n",
       "[185121 rows x 1 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Psuedoalignment for all samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This creates a list from 1 to 19 and adds 28, 29, 30\n",
    "samples = list(range(1, 20)) + [28, 29, 30] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0001_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0001_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 207,412,414 reads, 107,409,028 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 188.166\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,484 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0002_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0002_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 110,391,707 reads, 55,185,695 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 194.64\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,342 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0003_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0003_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 115,218,212 reads, 24,920,207 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 184.553\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,247 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0004_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0004_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 107,925,296 reads, 55,394,001 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 190.37\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,304 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0005_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0005_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 154,042,287 reads, 83,975,929 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 189.694\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,352 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0006_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0006_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 127,237,972 reads, 68,864,313 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 190.055\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,307 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0007_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0007_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 123,734,157 reads, 65,543,966 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 186.632\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,456 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0008_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0008_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 127,245,780 reads, 65,157,728 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 186.37\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,425 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0009_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0009_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 137,858,908 reads, 65,566,197 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 194.343\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,287 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0010_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0010_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 148,519,131 reads, 89,508,105 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 189.81\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,635 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0011_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0011_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 140,634,955 reads, 80,444,831 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 193.64\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,395 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0012_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0012_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 123,525,737 reads, 70,627,620 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 194.167\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,680 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0013_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0013_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 125,759,430 reads, 75,728,179 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 190.672\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,322 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0014_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0014_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 143,596,065 reads, 84,150,575 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 193.218\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 1,717 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 185,121\n",
      "[index] number of k-mers: 142,403,527\n",
      "[index] number of equivalence classes: 657,654\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /mnt/e/20231030 - 10051-NM/10051-NM-0015_S1_L005_R1_001.fastq.gz\n",
      "                             /mnt/e/20231030 - 10051-NM/10051-NM-0015_S1_L005_R2_001.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ..."
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "samples = list(range(1, 20)) + [28, 29, 30]  # Sample numbers\n",
    "\n",
    "for sample in samples:\n",
    "    sample_str = f\"{sample:04d}\"  # Formats the sample number as four digits\n",
    "\n",
    "    r1_path = f\"/mnt/e/20231030 - 10051-NM/10051-NM-{sample_str}_S1_L005_R1_001.fastq.gz\"\n",
    "    r2_path = f\"/mnt/e/20231030 - 10051-NM/10051-NM-{sample_str}_S1_L005_R2_001.fastq.gz\"\n",
    "\n",
    "    output_dir = f\"/mnt/e/kallisto/sample_{sample_str}\"\n",
    "\n",
    "    if os.path.exists(r1_path) and os.path.exists(r2_path):\n",
    "        # Run kallisto quant for each sample\n",
    "        !kallisto quant -i \"/mnt/e/fastq/GCF_000001405.40/ncbi_dataset/data/GCF_000001405.40/output_indexed.idx\" -o \"{output_dir}\" \"{r1_path}\" \"{r2_path}\"\n",
    "    else:\n",
    "        print(f\"Files for Sample {sample_str} not found, skipping...\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "samples = list(range(1, 20)) + [28, 29, 30]  # Sample numbers\n",
    "combined_df = pd.DataFrame()\n",
    "\n",
    "for sample in samples:\n",
    "    sample_str = f\"{sample:04d}\"\n",
    "    abundance_path = f\"/mnt/e/kallisto/sample_{sample_str}/abundance.tsv\"\n",
    "    \n",
    "    if os.path.exists(abundance_path):\n",
    "        sample_df = pd.read_csv(abundance_path, sep='\\t')\n",
    "        sample_df = sample_df[['target_id', 'tpm']].rename(columns={'tpm': f'Sample_{sample_str}'})\n",
    "        sample_df.set_index('target_id', inplace=True)\n",
    "\n",
    "        if combined_df.empty:\n",
    "            combined_df = sample_df\n",
    "        else:\n",
    "            combined_df = combined_df.join(sample_df, how='outer')\n",
    "    else:\n",
    "        print(f\"Abundance file for Sample {sample_str} not found, skipping...\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df.to_csv('/mnt/e/kallisto/pseudoalign_df.csv')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
